File: library/src/test/java/me/proxer/library/api/info/EntryEndpointTest.java
package me.proxer.library.api.info;

import me.proxer.library.ProxerTest;
import me.proxer.library.api.ProxerException;
import me.proxer.library.entitiy.info.*;
import me.proxer.library.entitiy.list.IndustryCore;
import me.proxer.library.enums.*;
import okhttp3.mockwebserver.MockResponse;
import org.junit.Test;

import java.io.IOException;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Arrays;
import java.util.Collections;
import java.util.EnumSet;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * @author Ruben Gees
 */
public class EntryEndpointTest extends ProxerTest {

    private static final SimpleDateFormat DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd hh:mm:ss");

    @Test
    public void testDefault() throws IOException, ProxerException, ParseException {
        server.enqueue(new MockResponse().setBody(fromResource("entry.json")));

        final Entry result = api.info()
                .entry("1")
                .build()
                .execute();

        assertThat(result).isEqualTo(buildTestEntry());
    }

    @Test
    public void testPath() throws ProxerException, IOException, InterruptedException {
        server.enqueue(new MockResponse().setBody(fromResource("entry.json")));

        api.info().entry("1")
                .build()
                .execute();

        assertThat(server.takeRequest().getPath()).isEqualTo("/api/v1/info/fullentry?id=1");
    }

    private Entry buildTestEntry() throws ParseException {
        return new Entry("6174", "LuCu LuCu",
                EnumSet.of(Genre.COMEDY, Genre.FANTASY, Genre.SEINEN, Genre.SLICE_OF_LIFE),
                EnumSet.noneOf(FskConstraint.class),
                "Humans are a despicable lot, committing sin after sin, filling the endless boundaries "
                        + "of the underworld with tortured souls. Now, it would seem, Hell isn't so endless after all, "
                        + "and it has become dangerously close to filling, and then overflowing into the human realm. "
                        + "Princess Lucuha sees this imminent disaster and has a plan: save Hell by making humans "
                        + "decent again. Of course, the angels can't simply allow demons to roam freely on Earth, "
                        + "and they do their best to stop Lucu and her dastardly plans.",
                Medium.MANGASERIES, 90, MediaState.FINISHED, 7, 1, 134,
                Category.MANGA, License.NOT_LICENSED, false, Arrays.asList(
                new Synonym("12322", "6174", SynonymType.ORIGINAL, "LuCu LuCu"),
                new Synonym("12323", "6174", SynonymType.JAPANESE, "るくるく"),
                new Synonym("44662", "6174", SynonymType.ORIGINAL_ALTERNATIVE, "Lucu Lucu"),
                new Synonym("44663", "6174", SynonymType.ORIGINAL_ALTERNATIVE, "LuCuLuCu")
        ), Collections.singleton(MediaLanguage.ENGLISH), Arrays.asList(
                new EntrySeasonInfo("1061", 2002, Season.UNSPECIFIED_ALT),
                new EntrySeasonInfo("15776", 2009, Season.UNSPECIFIED_ALT)
        ), Arrays.asList(
                new EntryTranslatorGroup("215", "SnoopyCool", Country.ENGLAND),
                new EntryTranslatorGroup("294", "FoOlRulez", Country.ENGLAND)
        ), Collections.singletonList(
                new IndustryCore("19", "Kodansha", IndustryType.PUBLISHER, Country.JAPAN)
        ), Arrays.asList(
                new Tag("93", "2027", DATE_FORMAT.parse("2016-06-18 14:14:22"), false,
                        false, "Dämonen", "In diesem Werk kommen Dämonen vor."),
                new Tag("299", "2028", DATE_FORMAT.parse("2016-06-18 14:14:30"), false,
                        false, "Slapstick", "Situationskomik, kommt ohne Worte aus.")
        ));
    }
}

File: library/src/test/kotlin/me/proxer/library/api/info/EntryEndpointTest.kt
package me.proxer.library.api.info

import me.proxer.library.ProxerTest
import me.proxer.library.entity.info.AdaptionInfo
import me.proxer.library.entity.info.Entry
import me.proxer.library.entity.info.EntrySeasonInfo
import me.proxer.library.entity.info.EntryTranslatorGroup
import me.proxer.library.entity.info.InfoGenre
import me.proxer.library.entity.info.InfoTag
import me.proxer.library.entity.info.Synonym
import me.proxer.library.entity.list.IndustryCore
import me.proxer.library.enums.Category
import me.proxer.library.enums.Country
import me.proxer.library.enums.IndustryType
import me.proxer.library.enums.License
import me.proxer.library.enums.MediaLanguage
import me.proxer.library.enums.MediaState
import me.proxer.library.enums.Medium
import me.proxer.library.enums.Season
import me.proxer.library.enums.SynonymType
import me.proxer.library.runRequest
import me.proxer.library.toProxerDateTime
import org.amshove.kluent.shouldBeEqualTo
import org.junit.jupiter.api.Test

/**
 * @author Ruben Gees
 */
class EntryEndpointTest : ProxerTest() {

    private val expectedEntry = Entry(
        id = "6174", name = "LuCu LuCu", fskConstraints = setOf(),
        description = """
            |Humans are a despicable lot, committing sin after sin, filling the endless boundaries of the
            | underworld with tortured souls. Now, it would seem, Hell isn't so endless after all, and it has become
            | dangerously close to filling, and then overflowing into the human realm. Princess Lucuha sees this
            | imminent disaster and has a plan: save Hell by making humans decent again. Of course, the angels can't
            | simply allow demons to roam freely on Earth, and they do their best to stop Lucu and her
            | dastardly plans.
        """.trimMargin().replace("\n", ""),
        medium = Medium.MANGASERIES, episodeAmount = 90, state = MediaState.FINISHED, ratingSum = 7, ratingAmount = 1,
        clicks = 134, category = Category.MANGA, license = License.NOT_LICENSED,
        adaptionInfo = AdaptionInfo(id = "2793", name = "KissXsis", medium = Medium.MANGASERIES),
        isAgeRestricted = false,
        synonyms = listOf(
            Synonym(id = "12322", entryId = "6174", type = SynonymType.ORIGINAL, name = "LuCu LuCu"),
            Synonym(id = "12323", entryId = "6174", type = SynonymType.JAPANESE, name = "るくるく"),
            Synonym(id = "44662", entryId = "6174", type = SynonymType.ORIGINAL_ALTERNATIVE, name = "Lucu Lucu"),
            Synonym(id = "44663", entryId = "6174", type = SynonymType.ORIGINAL_ALTERNATIVE, name = "LuCuLuCu")
        ),
        languages = setOf(MediaLanguage.ENGLISH),
        seasons = listOf(
            EntrySeasonInfo(id = "1061", year = 2002, season = Season.UNSPECIFIED),
            EntrySeasonInfo(id = "15776", year = 2009, season = Season.UNSPECIFIED)
        ),
        translatorGroups = listOf(
            EntryTranslatorGroup(id = "215", name = "SnoopyCool", country = Country.ENGLAND),
            EntryTranslatorGroup(id = "294", name = "FoOlRulez", country = Country.ENGLAND)
        ),
        industries = listOf(
            IndustryCore(id = "19", name = "Kodansha", type = IndustryType.PUBLISHER, country = Country.JAPAN)
        ),
        tags = listOf(
            InfoTag(
                id = "2027",
                entryTagId = "93",
                date = "2016-06-18 14:14:22".toProxerDateTime(),
                isRated = false,
                isSpoiler = false,
                name = "Dämonen",
                description = "In diesem Werk kommen Dämonen vor."
            ),
            InfoTag(
                id = "2028",
                entryTagId = "299",
                date = "2016-06-18 14:14:30".toProxerDateTime(),
                isRated = false,
                isSpoiler = false,
                name = "Slapstick",
                description = "Situationskomik, kommt ohne Worte aus."
            )
        ),
        genres = listOf(
            InfoGenre(
                id = "120694",
                entryTagId = "175",
                date = "2018-03-07 23:17:44".toProxerDateTime(),
                name = "Action",
                description = """
                    Dynamische Szenen, spannende Wettkämpfe und beeindruckende Kampfszenen prägen dieses Genre.
                """.trimIndent().replace("\n", "")
            ),
            InfoGenre(
                id = "120695",
                entryTagId = "174",
                date = "2018-03-07 23:17:44".toProxerDateTime(),
                name = "Adventure",
                description = """
                    |Es handelt sich meist um eine Geschichte über eine Reise oder Suche.
                    | Kurzum ein Abenteuer, das es zu bestehen gilt.
                """.trimMargin().replace("\n", "")
            )
        )
    )

    @Test
    fun testDefault() {
        val (result, _) = server.runRequest("entry.json") {
            api.info
                .entry("1")
                .build()
                .execute()
        }

        result shouldBeEqualTo expectedEntry
    }

    @Test
    fun testPath() {
        val (_, request) = server.runRequest("entry.json") {
            api.info.entry("1")
                .build()
                .execute()
        }

        request.path shouldBeEqualTo "/api/v1/info/fullentry?id=1"
    }
}
